In [1]:
from keras.datasets import mnist

# loading the dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


In [2]:
# let's print the shape of the dataset
print("X_train shape", X_train.shape)
print("y_train shape", y_train.shape)
print("X_test shape", X_test.shape)
print("y_test shape", y_test.shape)

X_train shape (60000, 28, 28)
y_train shape (60000,)
X_test shape (10000, 28, 28)
y_test shape (10000,)


Before we train a CNN model, let’s build a basic Fully Connected Neural Network for the dataset. The basic steps to build an image classification model using a neural network are:

* Flatten the input image dimensions to 1D (width pixels x height pixels)
* Normalize the image pixel values (divide by 255)
* One-Hot Encode the categorical column
* Build a model architecture (Sequential) with Dense layers
* Train the model and make predictions

In [3]:
# keras imports for the dataset and building our neural network
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D
from keras.utils import np_utils

In [4]:
# Flattening the images from the 28x28 pixels to 1D 787 pixels
X_train = X_train.reshape(60000, 784)
X_test = X_test.reshape(10000, 784)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [5]:
# normalizing the data to help with the training
X_train /= 255
X_test /= 255

In [6]:
# one-hot encoding using keras' numpy-related utilities
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)


In [7]:
# building a linear stack of layers with the sequential model
model = Sequential()
# hidden layer
model.add(Dense(100, input_shape=(784,), activation='relu'))
# output layer
model.add(Dense(10, activation='softmax'))

# looking at the model summary
model.summary()

# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 100)               78500     
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1010      
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


In [8]:
# training the model for 10 epochs
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test))

Epoch 1/10
469/469 [==============================] - 18s 5ms/step - loss: 0.6487 - accuracy: 0.8248 - val_loss: 0.2186 - val_accuracy: 0.9356
Epoch 2/10
469/469 [==============================] - 2s 4ms/step - loss: 0.1940 - accuracy: 0.9454 - val_loss: 0.1455 - val_accuracy: 0.9581
Epoch 3/10
469/469 [==============================] - 2s 4ms/step - loss: 0.1320 - accuracy: 0.9623 - val_loss: 0.1246 - val_accuracy: 0.9623
Epoch 4/10
469/469 [==============================] - 2s 4ms/step - loss: 0.1026 - accuracy: 0.9706 - val_loss: 0.1032 - val_accuracy: 0.9674
Epoch 5/10
469/469 [==============================] - 2s 4ms/step - loss: 0.0827 - accuracy: 0.9769 - val_loss: 0.0924 - val_accuracy: 0.9708
Epoch 6/10
469/469 [==============================] - 2s 4ms/step - loss: 0.0674 - accuracy: 0.9813 - val_loss: 0.0870 - val_accuracy: 0.9729
Epoch 7/10
469/469 [==============================] - 2s 4ms/step - loss: 0.0574 - accuracy: 0.9838 - val_loss: 0.0834 - val_accuracy: 0.9743
Epoch

We are getting a good validation accuracy of around 97% easily.

Let’s modify the above code to build a CNN model.

In [9]:
# keras imports for the dataset and building our neural network
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, Flatten
from keras.utils import np_utils

# to calculate accuracy
from sklearn.metrics import accuracy_score

# loading the dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# building the input vector from the 28x28 pixels
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1)
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# normalizing the data to help with the training
X_train /= 255
X_test /= 255

# one-hot encoding using keras' numpy-related utilities
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

# building a linear stack of layers with the sequential model
model = Sequential()
# convolutional layer
model.add(Conv2D(25, kernel_size=(3,3), strides=(1,1), padding='valid', activation='relu', input_shape=(28,28,1)))
model.add(MaxPool2D(pool_size=(1,1)))
# flatten output of conv
model.add(Flatten())
# hidden layer
model.add(Dense(100, activation='relu'))
# output layer
model.add(Dense(10, activation='softmax'))

# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

# training the model for 10 epochs
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test))

Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)
Epoch 1/10
469/469 [==============================] - 32s 9ms/step - loss: 0.3907 - accuracy: 0.8864 - val_loss: 0.0759 - val_accuracy: 0.9760
Epoch 2/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0670 - accuracy: 0.9802 - val_loss: 0.0559 - val_accuracy: 0.9800
Epoch 3/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0360 - accuracy: 0.9888 - val_loss: 0.0483 - val_accuracy: 0.9844
Epoch 4/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0219 - accuracy: 0.9937 - val_loss: 0.0512 - val_accuracy: 0.9840
Epoch 5/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0143 - accuracy: 0.9964 - val_loss: 0.0515 - val_accuracy: 0.9835
Epoch 6/10
469/469 [==============================] - 4s 8ms/step - loss: 0.0103 - accuracy: 0.9971 - val_loss: 0.0573 - val_accuracy: 0.9825
Epoch 7/10
469/469 [==============================] - 4s 8ms/st

With a single covolution layer val_acuracy improved from 97 to 98%.